File: services/analyzer/app/schemas/analysis.py
"""Mirrors packages/schemas/src/analysis.ts on the TypeScript side — keep
the two in sync by hand; there are few enough fields that a codegen step
would be more ceremony than it's worth right now (tracked as a possible
follow-up in docs/LIMITATIONS.md)."""

from __future__ import annotations

from enum import StrEnum

from app.schemas.base import CamelModel
from app.schemas.finding import Finding
from app.schemas.graph import FundFlowGraph
from app.schemas.ir import ProjectIR


class AnalysisStatus(StrEnum):
    QUEUED = "queued"
    INTAKE = "intake"
    COMPILING = "compiling"
    ANALYZING = "analyzing"
    DETECTING = "detecting"
    VALIDATING = "validating"
    SCORING = "scoring"
    COMPLETE = "complete"
    FAILED = "failed"


class SeverityCounts(CamelModel):
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0


class FindingScoreBreakdown(CamelModel):
    """Per-finding factor values behind the analysis's overall FLARE score —
    see docs/RISK_METHODOLOGY.md. Exposed so the UI can show *why* a score
    is what it is, not just the final number."""

    finding_id: str
    severity_weight: float
    confidence: float
    reachability: str
    reachability_weight: float
    asset_exposure: float
    dependency_criticality: float
    recovery_offset: float
    validated: bool
    contribution: float


class AnalysisSummary(CamelModel):
    id: str
    origin: str = "live"
    project_name: str
    project_version: str = "uploaded"
    status: AnalysisStatus
    created_at: str
    flare_score: float | None = None
    risk_band: str | None = None
    formula_version: str | None = None
    coverage: float | None = None
    coverage_notes: list[str] = []
    score_breakdown: list[FindingScoreBreakdown] = []
    finding_count: int = 0
    severity_counts: SeverityCounts = SeverityCounts()
    error: str | None = None
    ir: ProjectIR | None = None
    findings: list[Finding] = []
    graph: FundFlowGraph | None = None

File: packages/schemas/src/analysis.ts
import { z } from "zod";
import { FundFlowGraph } from "@flare/graph";
import { TaxonomyCategoryId } from "./taxonomy";

export const Severity = z.enum(["critical", "high", "medium", "low"]);
export type Severity = z.infer<typeof Severity>;

export const RiskBand = z.enum(["critical", "high", "medium", "low", "minimal"]);
export type RiskBand = z.infer<typeof RiskBand>;

export const AnalysisStatus = z.enum([
  "queued",
  "intake",
  "compiling",
  "analyzing",
  "detecting",
  "validating",
  "scoring",
  "complete",
  "failed",
]);
export type AnalysisStatus = z.infer<typeof AnalysisStatus>;

export const DataOrigin = z.enum(["demo", "live"]);
export type DataOrigin = z.infer<typeof DataOrigin>;

export const Reachability = z.enum(["public", "privileged", "theoretical"]);
export type Reachability = z.infer<typeof Reachability>;

export const Finding = z.object({
  id: z.string(),
  detectorId: z.string(),
  detectorVersion: z.string(),
  title: z.string(),
  taxonomy: TaxonomyCategoryId,
  severity: Severity,
  confidence: z.number().min(0).max(1),
  reachability: Reachability.default("theoretical"),
  file: z.string(),
  lineStart: z.number().int(),
  lineEnd: z.number().int(),
  codeExcerpt: z.string(),
  explanation: z.string(),
  triggeringCondition: z.string(),
  assetLockConsequence: z.string(),
  remediation: z.string(),
  relatedIncidentIds: z.array(z.string()).default([]),
  validated: z.boolean(),
  falsePositiveSuppressed: z.boolean().default(false),
  suppressionReason: z.string().optional(),
});
export type Finding = z.infer<typeof Finding>;

export const SeverityCounts = z.object({
  critical: z.number().int().default(0),
  high: z.number().int().default(0),
  medium: z.number().int().default(0),
  low: z.number().int().default(0),
});
export type SeverityCounts = z.infer<typeof SeverityCounts>;

export const FindingScoreBreakdown = z.object({
  findingId: z.string(),
  severityWeight: z.number(),
  confidence: z.number(),
  reachability: Reachability,
  reachabilityWeight: z.number(),
  assetExposure: z.number(),
  dependencyCriticality: z.number(),
  recoveryOffset: z.number(),
  validated: z.boolean(),
  contribution: z.number(),
});
export type FindingScoreBreakdown = z.infer<typeof FindingScoreBreakdown>;

export const AnalysisSummary = z.object({
  id: z.string(),
  origin: DataOrigin,
  projectName: z.string(),
  projectVersion: z.string(),
  status: AnalysisStatus,
  createdAt: z.string(),
  flareScore: z.number().min(0).max(100).nullable(),
  riskBand: RiskBand.nullable(),
  formulaVersion: z.string().nullable().default(null),
  coverage: z.number().min(0).max(1).nullable(),
  coverageNotes: z.array(z.string()).default([]),
  scoreBreakdown: z.array(FindingScoreBreakdown).default([]),
  findingCount: z.number().int(),
  severityCounts: SeverityCounts,
  error: z.string().nullable().default(null),
  findings: z.array(Finding).default([]),
  graph: FundFlowGraph.nullable().default(null),
});
export type AnalysisSummary = z.infer<typeof AnalysisSummary>;
